nixos/modules/services/databases/openldap.nix

```diff
@@ -244,57 +356,156 @@ in
   };
 
   meta = {
-    maintainers = [ lib.maintainers.mic92 ];
+    maintainers = with lib.maintainters; [ mic92 kwohlfahrt ];
   };
 
-
-  ###### implementation
-
   config = mkIf cfg.enable {
-    assertions = [
-      {
-        assertion = cfg.configDir != null || cfg.rootpwFile != null || cfg.rootpw != null;
-        message = "services.openldap: Unless configDir is set, either rootpw or rootpwFile must be set";
-      }
-    ];
+    warnings = let
+      deprecations = [
+        { old = "logLevel"; new = "attrs.olcLogLevel"; }
+        { old = "defaultSchemas";
+          new = "children.\"cn=schema\".includes";
+          newValue = "[\n    ${lib.concatStringsSep "\n    " [
+            "\${pkgs.openldap}/etc/schema/core.ldif"
+            "\${pkgs.openldap}/etc/schema/cosine.ldif"
+            "\${pkgs.openldap}/etc/schema/inetorgperson.ldif"
+            "\${pkgs.openldap}/etc/schema/nis.ldif"
+          ]}\n  ]"; }
+        { old = "database"; new = "children.\"cn={1}${cfg.database}\""; newValue = "{ }"; }
+        { old = "suffix"; new = "children.\"cn={1}${cfg.database}\".attrs.suffix"; }
+        { old = "dataDir"; new = "children.\"cn={1}${cfg.database}\".attrs.olcDbDirectory"; }
+        { old = "rootdn"; new = "children.\"cn={1}${cfg.database}\".attrs.olcRootDN"; }
+        { old = "rootpw"; new = "children.\"cn={1}${cfg.database}\".attrs.olcRootPW"; }
+        { old = "rootpwFile";
+          new = "children.\"cn={1}${cfg.database}\".attrs.olcRootPW";
+          newValue = "{ path = \"${cfg.rootpwFile}\"; }";
+          note = "The file should contain only the password (without \"rootpw \" as before)"; }
+      ];
+    in (optional (cfg.extraConfig != "" || cfg.extraDatabaseConfig != "") ''
+      The options `extraConfig` and `extraDatabaseConfig` of `services.openldap`
+      are deprecated. This is due to the deprecation of `slapd.conf`
+      upstream. Please migrate to `services.openldap.settings`.
+
+      You can run:
+        slapcat -F ${configDir} -n0 -H 'ldap:///???(!(objectClass=olcSchemaConfig))'
```

Comment:
I get:

```
$ slapcat -F /etc/openldap/slapd.d -n0 -H 'ldap:///???(!(objectClass=olcSchemaConfig))'
5f395c1f invalid config directory /etc/openldap/slapd.d, error 2
```

I think we should include that you need to deploy your old configuration at least once before the directory is created?

Reply:
Using this configuration: https://github.com/Mic92/dotfiles/blob/master/nixos/eve/modules/openldap/default.nix#L19

it also seems to not pick up yet my schema:

● openldap.service - LDAP server
     Loaded: loaded (/nix/store/a0vf2f37b63h1cziws81ml9f4m0v4cv2-unit-openldap.service/openldap.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Sun 2020-08-16 16:41:32 UTC; 6min ago
    Process: 6127 ExecStartPre=/nix/store/0fyrxmxa7ibkca2b4n3pjwfr536n3579-unit-script-openldap-pre-start/bin/openldap-pre-start (code=exited, status=1/FAILURE)
         IP: 0B in, 0B out
        CPU: 36ms

Aug 16 16:41:32 eve systemd[1]: Starting LDAP server...
Aug 16 16:41:32 eve openldap-pre-start[6184]: config file testing succeeded
Aug 16 16:41:32 eve slaptest[6184]: DIGEST-MD5 common mech free
Aug 16 16:41:32 eve openldap-pre-start[6306]: 5f3961bc olcObjectClasses: value #0 olcObjectClasses: ObjectClass not found: "uidObject"
Aug 16 16:41:32 eve openldap-pre-start[6306]: 5f3961bc config error processing cn=config: olcObjectClasses: ObjectClass not found: "uidObject"
Aug 16 16:41:32 eve openldap-pre-start[6306]: slaptest: bad configuration directory!
Aug 16 16:41:32 eve systemd[1]: openldap.service: Control process exited, code=exited, status=1/FAILURE
Aug 16 16:41:32 eve systemd[1]: openldap.service: Failed with result 'exit-code'.
Aug 16 16:41:32 eve systemd[1]: Failed to start LDAP server.

Reply:
Yes, good point.